Add sale_payment_term_interest

guewen opened the pull request:
This module allows to compute interest fees based on payment terms and
to add a line in the sales orders with the computed amount.

Commit:
Started account_invoice_interest: initial files

Commit:
Add a method on payment terms to compute the interests

Commit:
Ad an invoice line with the interest amount

Commit:
Check interest line amount before to create the move

If the interest amount is wrong, ask to update it.

Commit:
Add a comment for a limitation

Commit:
onchange to automatically update the interest line

When the payment term is changed. The button to update it it still needed
because a change on the lines can't update the line (limitation of the
onchanges).

Commit:
Put the line at the very end of the lines.

Commit:
Interests are now added on sale order

Rename the module and the files according to this change.

Commit:
Remove the onchange: the total is updated at save

So the onchange does not work. The button should be used

Commit:
Remove the interest line including its taxes to recompute the interest

Commit:
Improve display of the (update interests) button

Commit:
Automatically update the interest line on writes

Commit:
Do not override the line's description on updates

Commit:
Add tests for sale_order_interest

Commit:
Documented the readme

Commit:
Document the field interest_rate

Commit:
Rename module to sale_payment_term_interest

Commit:
Use english in the product

Commit:
Add translation template

Commit:
Add french translation

Commit:
Revert "Add translation template"

This reverts commit 4fbdc1f56d7efd724ca83076a27771e83f364770.

dufresnedavid commented on sale_payment_term_interest/model/account_payment_term.py:
```diff
@@ -0,0 +1,89 @@
+                line_amount = round(value * line.value_amount,
+                                    precision_digits=prec)
+            elif line.value == 'balance':
+                line_amount = round(amount, prec)
```
Why do you change the digit precision? It would seem to me more logical to let the precision dp.get_precision('Payment Term').

guewen replied:
This code is the same than `account_payment_term.compute()` in the core account module. Since I want interests based on the amounts due for each term, I have to use the exact same computation than the `compute()` method.

dufresnedavid commented on sale_payment_term_interest/tests/test_sale_order.py:
```diff
@@ -0,0 +1,157 @@
+            'product_uom': product1.uom_id.id,
+            'price_unit': 100,
+            'tax_id': [(6, 0, [tax.id])],
+        }
```
The beggining of each test is almost the same here in the 3 cases. You could move what is repeated to setUp and then just write in your objects what changes from one unit test to the other.

ghost commented:
@guewen Added you a few comments. Otherwise, looks good to me.

Commit:
Error when called on several ids

Commit:
Remove repetition in tests

Commit:
Add a comment in the compute_interest method

Explaining why the precision is 'Account'

guewen commented:
> @guewen Added you a few comments. Otherwise, looks good to me.

Thanks for reviewing. I removed the repetition in the tests and answered you for the decimal precision, also added a comment in the code.

gurneyalex commented on sale_payment_term_interest/README.rst:
```diff
@@ -0,0 +1,54 @@
+
+ * Go to *Invoicing > Configuration > Payment Terms*
+ * On the lines of the payment terms, you can configure an interest
+   rate. It is based on the number of days.
```
I find this strange, but I'm don't know the general practice here. Most of the time I see an annual interest rate, which is divided by 12 to get a monthly rate (and then by 30 to get a daily rate if needed).

Is this meant to be used in a B2B or a B2C context?

dufresnedavid replied:
I agress with @gurnalex here. Interest is in most cases computed on an annual basis. But, dividing by 12 for monthly is not the way to do this, because interest is rarely compounded annually.

guewen replied:
Done, the rate is now on an annual basis.

gurneyalex commented on sale_payment_term_interest/tests/test_payment_term.py:
```diff
@@ -0,0 +1,61 @@
+            'value': 'procent',
+            'value_amount': 0.3333,
+            'days': 40,
+            'interest_rate': 0.041666,
```
for readability purpose, it would be nicer in my opinion to have this written as 15/(12*30) (and to add `from __future__ import division`) at the beginning of the file (and I still feel it would be more user friendly to have 15 for the value here).

guewen replied:
Done

gurneyalex commented on sale_payment_term_interest/tests/test_payment_term.py:
```diff
@@ -0,0 +1,61 @@
+        self.assertEquals(len(terms), 3)
+        terms = sorted(terms, key=itemgetter(0))
+        line1, line2, line3 = terms
+        self.assertAlmostEqual(line1[1], 156.43)
```
same as above: if you put in the mathematical expression instead of the result, you get a test that will be easier to maintain.

guewen replied:
Done

Commit:
The interest is now an annual rate

Commit:
Update translations

ghost commented:
:+1:

yvaucher commented:
:+1:  it says in https://github.com/OCA/sale-workflow/pull/123 that this one superseeds it. However #123 is merged. Was it intentional?

guewen commented:
> it says in #123 that this one superseeds it. However #123 is merged. Was it intentional?

#123 has been reverted

yvaucher commented:
@guewen Ok thanks I proceed with the merge then
